opt: generate lookup joins on partitioned indexes

mgartner opened the pull request:
#### xform: add rule expectations for GenerateLookupJoins tests

Release note: None

#### opt: generate lookup join for column constrained to multiple constants

Previously, the optimizer could create lookup join keys from filters
that constrain a column to a single constant value. This was done by
wrapping the join input in a Project that projected the constant value,
and using this new column as a key column.

This commit generalizes this behavior so that lookup join keys can also
be created from filters that constrain a column to multiple, non-ranging
constant values. The constant values are cross-joined with the input,
and the joined column is used as a key column. If a column is
constrained to a single constant value, the cross join normalizes to a
Project identical to the Projects constructed prior to this commit.

Release note (performance improvement): The query optimizer can use
filters that constrained columns to multiple constant values to generate
lookup joins. For example, a join filter `x.a = y.a AND y.b IN (1, 2)`
can be used to generate a lookup join on table `y` assuming that it has
an index on `(a, b)` or `(b, a)`.

#### opt: generate lookup joins with CHECK constraints and computed columns

Previously, only explicit filters were used to generated lookup join key
columns. Now lookup join keys can be generated from CHECK constraints
and computed column expressions.

With this commit and the previous commit, lookup joins on partitioned
indexes are explored by the optimizer.

Release note (performance improvement): The query optimizer now explores
plans with lookup joins on partitioned indexes, resulting in more
efficient query plans in some cases.

#### xform: do not propagate join hints to GenerateLookupJoins cross joins

This commit fixes a bug that prevented a `LOOKUP` join hint from
producing a plan with a lookup join. Previously, the hint was propagated
to the synthesized cross join created as input to the lookup join. This
artificially inflated the cost of the cross join, making the lookup join
too costly to be selected as the optimal plan.

Release note: None

mgartner committed:
xform: add rule expectations for GenerateLookupJoins tests

Release note: None

mgartner committed:
opt: generate lookup join for column constrained to multiple constants

Previously, the optimizer could create lookup join keys from filters
that constrain a column to a single constant value. This was done by
wrapping the join input in a Project that projected the constant value,
and using this new column as a key column.

This commit generalizes this behavior so that lookup join keys can also
be created from filters that constrain a column to multiple, non-ranging
constant values. The constant values are cross-joined with the input,
and the joined column is used as a key column. If a column is
constrained to a single constant value, the cross join normalizes to a
Project identical to the Projects constructed prior to this commit.

Release note (performance improvement): The query optimizer can use
filters that constrained columns to multiple constant values to generate
lookup joins. For example, a join filter `x.a = y.a AND y.b IN (1, 2)`
can be used to generate a lookup join on table `y` assuming that it has
an index on `(a, b)` or `(b, a)`.

mgartner committed:
opt: generate lookup joins with CHECK constraints and computed columns

Previously, only explicit filters were used to generated lookup join key
columns. Now lookup join keys can be generated from CHECK constraints
and computed column expressions.

With this commit and the previous commit, lookup joins on partitioned
indexes are explored by the optimizer.

Release note (performance improvement): The query optimizer now explores
plans with lookup joins on partitioned indexes, resulting in more
efficient query plans in some cases.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57690)
<!-- Reviewable:end -->

mgartner commented on pkg/sql/opt/xform/testdata/external/trading:
```diff
@@ -1110,27 +1110,41 @@ limit
  │         ├── project
  │         │    ├── columns: quantity:14!null accountname:10!null
- │         │    ├── inner-join (hash)
+ │         │    ├── inner-join (lookup inventorydetails)
```
This new query plan is better I think. Previously it was scanning `inventorydetails` (a large table) and only constraining by `accountname` (low cardinality). The new plan is a lookup join which is a "pseudo-scan" constraining  `accountname`, `dealerid`, and `cardid`. What do you think @andy-kimball?

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/xform/testdata/rules/join, line 3011 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO33-D91jA1FafJnzvl:-MO33-D91jA1FafJnzvm:b-ai4lcg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a5ab246a1c6b254f73037648ec3bc6142c4603/pkg/sql/opt/xform/testdata/rules/join#L3011)):*
> ```
>  │    │    │    ├── (10,)
>  │    │    │    ├── (20,)
>  │    │    │    └── (30,)
> ```

This is interesting. What do we currently do with geo-partitioned regular indices when the input rows do not supply the partition column values and the values are injected using a predicate like the `b IN (10, 20, 30)` here. Does this mean we were unable to use the partitioned index currently?

Will this cross-join with a constant table also be how we inject these constants into the input row for feeding into the inverted join?

Do we know enough (in the optimizer) after doing this cross-join to ship the inputs prefixed by 10, 20, 30 to the relevant regions (when this is a partitioning field), so that the lookup can be local?
 


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/xform/testdata/rules/join, line 3011 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO33-D91jA1FafJnzvl:-MO3dqdHAeLptF11AAaK:b-mot0wz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a5ab246a1c6b254f73037648ec3bc6142c4603/pkg/sql/opt/xform/testdata/rules/join#L3011)):*
> Does this mean we were unable to use the partitioned index currently?

Yes. Prior to this PR, we use partitioned indexes for other expressions like scans, but not for lookup joins.

> Will this cross-join with a constant table also be how we inject these constants into the input row for feeding into the inverted join?

Yes, that's the plan. I had started work on inverted joins on partitioned inverted indexes, and realized that an inverted join is similar to a lookup join on a non-inverted index. Once I realized that lookup joins did not utilize partitioned indexes, I decided to start with that as a "warm-up" to the inverted join work.

> Do we know enough (in the optimizer) after doing this cross-join to ship the inputs prefixed by 10, 20, 30 to the relevant regions (when this is a partitioning field), so that the lookup can be local?

I'm not 100% sure, but I think this information is available at the `distsql` level but not in the optimizer. @RaduBerinde or @rytaft may be able to answer this better.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Very cool!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-:-MO49wYC-9o6Ge4dM-lx:b-kkp8rn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/xform/testdata/rules/join, line 3295 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO49cSBCQT7OfGa9KMQ:-MO49cSBCQT7OfGa9KMR:bdjr81z) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a5ab246a1c6b254f73037648ec3bc6142c4603/pkg/sql/opt/xform/testdata/rules/join#L3295)):*
> ```
> ----
> 
> opt expect=GenerateLookupJoinsWithFilter
> ```

[nit] maybe explain that the inlining rule converts the single-left row cross join with project


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r1, 3 of 5 files at r2, 12 of 12 files at r4, 11 of 11 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-:-MO6oTxD2P6etvFVKfAu:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/xform/testdata/rules/join, line 3011 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO33-D91jA1FafJnzvl:-MO6fT85AcaTsitEY8iv:btl0fri) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a5ab246a1c6b254f73037648ec3bc6142c4603/pkg/sql/opt/xform/testdata/rules/join#L3011)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

> Does this mean we were unable to use the partitioned index currently?

Yes. Prior to this PR, we use partitioned indexes for other expressions like scans, but not for lookup joins.

> Will this cross-join with a constant table also be how we inject these constants into the input row for feeding into the inverted join?

Yes, that's the plan. I had started work on inverted joins on partitioned inverted indexes, and realized that an inverted join is similar to a lookup join on a non-inverted index. Once I realized that lookup joins did not utilize partitioned indexes, I decided to start with that as a "warm-up" to the inverted join work.

> Do we know enough (in the optimizer) after doing this cross-join to ship the inputs prefixed by 10, 20, 30 to the relevant regions (when this is a partitioning field), so that the lookup can be local?

I'm not 100% sure, but I think this information is available at the `distsql` level but not in the optimizer. @RaduBerinde or @rytaft may be able to answer this better.

</blockquote></details>

@mgartner is right that this would happen at the DistSQL planner level, not at the optimizer level. But I'm pretty sure that what you're describing is not how lookup joins are planned, even by the DistSQL planner. Although @raduberinde or someone from SQL Execution would know better here.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/xform/testdata/rules/join, line 3011 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO33-D91jA1FafJnzvl:-MO6sHvRB5dtUQDom3cR:b-85c9t9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a5ab246a1c6b254f73037648ec3bc6142c4603/pkg/sql/opt/xform/testdata/rules/join#L3011)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

@mgartner is right that this would happen at the DistSQL planner level, not at the optimizer level. But I'm pretty sure that what you're describing is not how lookup joins are planned, even by the DistSQL planner. Although @raduberinde or someone from SQL Execution would know better here.
</blockquote></details>

I think the lookup joins are currently distributed based on the locations the input happens to be generated at. If the input was already geo-partitioned, that seems fine. If we've generated the partitioning columns using the cross-join, like this example, there may be cases where "moving" the input to the relevant regions may be cheaper. For example, if the input is small, but the output of the join is large and is subsequently aggregated.

I am not suggesting any changes for this PR -- just curious about how query execution placement in a multi-region setting will behave.


<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/opt/xform/testdata/rules/join, line 3011 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO33-D91jA1FafJnzvl:-MO7-DAF6bf-hwjH6eOR:b-1ok172) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a5ab246a1c6b254f73037648ec3bc6142c4603/pkg/sql/opt/xform/testdata/rules/join#L3011)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think the lookup joins are currently distributed based on the locations the input happens to be generated at. If the input was already geo-partitioned, that seems fine. If we've generated the partitioning columns using the cross-join, like this example, there may be cases where "moving" the input to the relevant regions may be cheaper. For example, if the input is small, but the output of the join is large and is subsequently aggregated.

I am not suggesting any changes for this PR -- just curious about how query execution placement in a multi-region setting will behave.
</blockquote></details>

Yea, that's a good point. We'll have to see if this is a common use case, in which case it may be worth doing in the DistSQL planner. It seems non-trivial, though. Also, this might be something we can plan in the optimizer some day (we are eventually hoping to make some of the distribution decisions inside the optimizer).


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/xform/testdata/rules/join, line 3295 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO49cSBCQT7OfGa9KMQ:-MO7eq_B6rDoFT8p3dWe:b-k56qnu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a5ab246a1c6b254f73037648ec3bc6142c4603/pkg/sql/opt/xform/testdata/rules/join#L3295)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] maybe explain that the inlining rule converts the single-left row cross join with project
</blockquote></details>

This happens in quite a few tests in this file, so instead of commenting on each one, I added some commentary to the `GenerateLookupJoins` function.


<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/opt/xform/join_funcs.go, line 306 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO858o3A21HKFC3_d87:-MO858o3A21HKFC3_d88:b-f650li) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3fd7215c63ad823db01f89b38b0566caf532c3/pkg/sql/opt/xform/join_funcs.go#L306)):*
> ```Go
> 				},
> 			)
> 			lookupJoin.Input = c.e.f.ConstructInnerJoin(lookupJoin.Input, values, nil /* on */, joinPrivate)
> ```

Reminder from our discussion just now that this should probably not copy the join private


<!-- Sent from Reviewable.io -->

mgartner committed:
xform: do not propagate join hints to GenerateLookupJoins cross joins

This commit fixes a bug that prevented a `LOOKUP` join hint from
producing a plan with a lookup join. Previously, the hint was propagated
to the synthesized cross join created as input to the lookup join. This
artificially inflated the cost of the cross join, making the lookup join
too costly to be selected as the optimal plan.

Release note: None

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/xform/join_funcs.go, line 306 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO858o3A21HKFC3_d87:-MO8Nyl80DVgDiXe_dph:brrsex8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3fd7215c63ad823db01f89b38b0566caf532c3/pkg/sql/opt/xform/join_funcs.go#L306)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Reminder from our discussion just now that this should probably not copy the join private
</blockquote></details>

Thanks! @rytaft @raduberinde PTAL at the new commit.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/xform/join_funcs.go, line 306 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO858o3A21HKFC3_d87:-MO8OK8G1y123erPla-G:b-wuwaxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3fd7215c63ad823db01f89b38b0566caf532c3/pkg/sql/opt/xform/join_funcs.go#L306)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Thanks! @rytaft @raduberinde PTAL at the new commit.
</blockquote></details>

Also please confirm that my assumption that "reordering a cross-join is pointless". Or is there a benefit in a cross-join to having fewer rows in the left child than the right?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/xform/join_funcs.go, line 306 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO858o3A21HKFC3_d87:-MO8bO8o-ZKTrwb0uFZ-:b-wv21i9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3fd7215c63ad823db01f89b38b0566caf532c3/pkg/sql/opt/xform/join_funcs.go#L306)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Also please confirm that my assumption that "reordering a cross-join is pointless". Or is there a benefit in a cross-join to having fewer rows in the left child than the right?
</blockquote></details>

Well, we store the right side in memory completely (with fallback to disk if necessary). So if we're cross-joining 2 rows with 1M rows, we really want to store the small side.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/xform/join_funcs.go, line 306 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO858o3A21HKFC3_d87:-MO8cNsD015nJ9u4Chwm:bl3r1z1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3fd7215c63ad823db01f89b38b0566caf532c3/pkg/sql/opt/xform/join_funcs.go#L306)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Well, we store the right side in memory completely (with fallback to disk if necessary). So if we're cross-joining 2 rows with 1M rows, we really want to store the small side.
</blockquote></details>

If a lookup join is chosen as the optimal plan, its unlikely that either side of this child cross-join is very large. A lookup is performed for each output row of the cross-join, which would be costly if the cross-join produced many rows.

It seems sane to me to leave the constants on the right and not attempt reordering. The constants are a single column, so even in cases where there are more constants than input rows its plausible that the constants would have a smaller memory footprint.

But if you think its safer to permit reordering of this cross-join then I'm happy to. Also happy to try to benchmark the cost of this reordering vs no reordering if that helps us decide which direction to take.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/xform/join_funcs.go, line 306 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO858o3A21HKFC3_d87:-MO8fVJW39ke6GeCRMNP:b-i1cjtq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3fd7215c63ad823db01f89b38b0566caf532c3/pkg/sql/opt/xform/join_funcs.go#L306)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

If a lookup join is chosen as the optimal plan, its unlikely that either side of this child cross-join is very large. A lookup is performed for each output row of the cross-join, which would be costly if the cross-join produced many rows.

It seems sane to me to leave the constants on the right and not attempt reordering. The constants are a single column, so even in cases where there are more constants than input rows its plausible that the constants would have a smaller memory footprint.

But if you think its safer to permit reordering of this cross-join then I'm happy to. Also happy to try to benchmark the cost of this reordering vs no reordering if that helps us decide which direction to take.
</blockquote></details>

Having the constants on the right make sense, but why disable reordering? We should let other rules run as usual unless we have a strong reason to disable them.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 1 of 1 files at r6, 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde)

___
*[pkg/sql/opt/xform/join_funcs.go, line 306 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO858o3A21HKFC3_d87:-MO8gqpgCazw5fFQHh1V:b-b0b4g4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3fd7215c63ad823db01f89b38b0566caf532c3/pkg/sql/opt/xform/join_funcs.go#L306)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Having the constants on the right make sense, but why disable reordering? We should let other rules run as usual unless we have a strong reason to disable them.
</blockquote></details>

Yea, I'd also err on the side of just allowing reordering. I don't see a benefit in disabling it here. 

Might also be worth adding a test that using an `INNER LOOKUP JOIN` hint works (since using the INNER INVERTED JOIN hint was how we identified the issue with copying the join private in the other case).


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/xform/join_funcs.go, line 306 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-MO858o3A21HKFC3_d87:-MO8w8-_7PUacRgZUejD:breneey) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3fd7215c63ad823db01f89b38b0566caf532c3/pkg/sql/opt/xform/join_funcs.go#L306)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Yea, I'd also err on the side of just allowing reordering. I don't see a benefit in disabling it here. 

Might also be worth adding a test that using an `INNER LOOKUP JOIN` hint works (since using the INNER INVERTED JOIN hint was how we identified the issue with copying the join private in the other case).
</blockquote></details>

SGTM. I've added a test with `INNER LOOKUP JOIN` and explained why the join flags aren't propagated in a comment. And updated the commit message and PR description.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57690#-:-MO9B3SG58BK7LelRj5g:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+
